### ISSUE TITLE
`single_component_path_imports` warns about imports that are necessary

### ISSUE THREAD

**Issue by hawkw** [rust-lang/rust-clippy#5539]
The `single_component_path_imports` lint will warn about imports which are re-exported by a `pub use self`. For example, see [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=81424685518298ffe612253540062ffb), where an item from the `log` crate is re-exported from `self`:
```rust
use log;

pub use self::{log::Log, some_mod::SomeType};

mod some_mod {
    pub struct SomeType;
}
```
Clippy will emit this warning:
```
    Checking playground v0.0.1 (/playground)
warning: this import is redundant
 --> src/lib.rs:1:1
  |
1 | use log;
  | ^^^^^^^^ help: remove it entirely
  |
  = note: `#[warn(clippy::single_component_path_imports)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_component_path_imports

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.60s
```

However, since `log` is re-exported from `self`, this error is incorrect. If the `use log` statement is removed, the reexport can no longer be resolved (as in [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=af7effa92a8ec5aa17266f8f95334fdb)):

```rust
pub use self::{log::Log, some_mod::SomeType};

mod some_mod {
    pub struct SomeType;
}
```
Which fails to compile with this error:
```
   Compiling playground v0.0.1 (/playground)
error[E0432]: unresolved import `self::log`
 --> src/lib.rs:1:16
  |
1 | pub use self::{log::Log, some_mod::SomeType};
  |                ^^^ maybe a missing crate `log`?

error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.
error: could not compile `playground`.
```

This isn't a _huge_ deal, since the code can be changed to 
```rust
pub use log::Log;
pub use self::some_mod::SomeType;
```
which is arguably clearer anyway. However, it is a case where clippy will suggest a change that results in code not compiling, which isn't great.

Clippy version:
```
:; cargo clippy -V
clippy 0.0.212 (204bb9b 2020-03-17)
```

**Comment by matthiaskrgr** [comment]
Could be a duplicate of #5210?

**Comment by hawkw** [comment]
@matthiaskrgr whoops, yeah, this looks like a duplicate; I didn't see that issue. My bad!
